lib/govuk_publishing_components/presenters/select_with_search_helper.rb

```diff
@@ -0,0 +1,66 @@
+##
+# Helper methods for use with the "select-with-search" component.
+# This wraps and extends the Select component's helper methods to add support for option groups.
+##
+module GovukPublishingComponents
+  module Presenters
+    class SelectWithSearchHelper
+      include ActionView::Helpers::FormOptionsHelper
+
+      attr_reader :options, :selected_option
+
+      delegate :describedby,
+               :error_id,
+               :error_message,
+               :hint_id,
+               :hint,
+               :label_classes,
+               :select_classes,
+               to: :@select_helper
+
+      def initialize(local_assigns)
+        @select_helper = SelectHelper.new(local_assigns.except(:options, :grouped_options))
+        @options = local_assigns[:options]
+        @grouped_options = local_assigns[:grouped_options]
```

Comment:
I could go either way on this, but did you consider a single options parameter that calls the appropriate method based on structure?

Reply:
Yeah – I did try that out first, as it was my initial preference too.

We essentially just massage the data a little before passing it through to the Rails helpers [options_for_select] or [grouped_options_for_select].

I was originally expecting to be able to just sniff the type of options object passed in – e.g. an Array for 'flat' options, or a Hash for grouped options – but I soon realised it wasn't as simple as that. It turns out [grouped_options_for_select] accepts both Array and Hash inputs.

So without going too far into the detail and sniffing out different types of Array structure, I ended up just re-using the same approach that Rails itself uses – namely, different methods/parameters to differentiate between 'flat' vs grouped lists of options. That felt like the easiest and most conventional way to solve the problem.

> I could go either way on this

I feel the same. I don't have a strong feeling either way, so am happy to change it if you think the other approach would be better.

[options_for_select]: https://apidock.com/rails/ActionView/Helpers/FormOptionsHelper/options_for_select
[grouped_options_for_select]: https://apidock.com/rails/v5.2.3/ActionView/Helpers/FormOptionsHelper/grouped_options_for_select

Reply:
Sounds good, that makes sense, just a shame that rails strayed from HTML conventions on this occasion.